docs/guides/contexts_expressions.md

```diff
@@ -0,0 +1,390 @@
+# Contexts and Expressions in napari
+
+```{note}
+This page is mostly aimed at developers who are interested in contributing to
+or understanding the inner workings of napari.  While the concept of expressions
+*will* be relevant for plugin developers writing plugin manifests, it won't be
+necessary for plugin devs or end-users to understand the implementation details
+described on this page.
+```
+
+*Contexts* and *Expressions* are two concepts being introduced along with the
+second-generation napari plugin engine (npe2) that capture the abstract idea of
+"some condition" (an `Expression`) that can be evaluated at some later time,
+with a concrete set of keys and values (the `Context`).
+
+## Python expressions
+
+In Python, **expressions** are simple combinations of **values** and
+**operations** that can be reduced to a single value. For example, `1 > 5` is an
+expression that always reduces to the value `False` when evaluated.  `x > 5 and
+y == 'hello'` is also an expression that reduces to a boolean value; however, in
+order to evaluate that expression, we need to be able to fill in the values for
+the variable **names** "`x`" and "`y`". Those values are provided by some
+**context** (or "namespace"), which maps the variable names to their values.
+
+The value of an `expression` depends on the context in which it is evaluated.
+
+```python
+In [1]: expression = "x > 5 and y == 'hello'"
+
+In [2]: context_a = {'x': 7, 'y': 'hello'}
+
+In [3]: context_b = {'x': 8, 'y': 'howdie!'}
+
+In [4]: eval(expression, context_a)
+Out[4]: True
+
+In [5]: eval(expression, context_b)
+Out[5]: False
+```
+
+## Napari expressions
+
+In napari, we'd like to be able to capture the concept of some condition being
+`True` or `False`, *prior* to actually having the context required to evaluate
+it.  For example, a plugin (or napari itself) might want to stipulate that a
+given function should only be enabled when "the active layer has at least 3
+dimensions".
+
+At runtime, in Python code, this might be captured by the expression:
+
+```python
+viewer.layers.selection.active.data.ndim >= 3
+```
+
+However, if you don't have access to the future `viewer` instance, that
+doesn't work.  So napari has the concept of `Expr` objects that represent an
+expression "without a context", to be evaluated later.
```

Comment:
I really like this problem statement and I wonder if it shouldn't go right at the top of the guide.